docs/blueprints/EP035.rst

```diff
@@ -0,0 +1,231 @@
+
+##########
+Motivation
+##########
+
+In order for NApps which produce service interruptions
+to receive information on what services they are affecting,
+there needs to be a two way communication channel between
+the service provider and the interruption source.
+The kytos event bus provides only one way communication,
+making it unsuitable for this task.
+
+#############
+Specification
+#############
+
+The following is a set of proposed changes for
+how to handle interruptions in the future,
+in order to have visibility on there causes and effects.
+
+Kytos Core Changes
+##################
+
+In order for services to identify which interruptions are affecting the devices
+they use, a mechanism needs to be added to the device representation for
+identify interruption causes.
+
+
+Extend ``status_funcs`` to produce a ``status_reason``
+------------------------------------------------------
+
+Currently, when an object is queried about its ``status``,
+the reason for that status is unknown.
+What we propose here is adding a new property to
+``GenericEntity`` called ``status_reason``,
+and adding in ``status_reason_func``
+to produce values for ``status_reason``.
+The produced ``status_reason`` will be a set of strings,
+so all possible reasons a device might be disable
+can be easily represented.
+
+Example usage:
+
+.. code-block:: python3
+
+    from kytos.core.switch import Switch
+    maintenance_switches = set()
+
+    def switch_maintenance_status(switch):
+        if switch.id in maintenance_switches:
+            return EntityStatus.DOWN
+        return EntityStatus.UP
+
+    Switch.register_status_func('maintenance', switch_maintenance_status)
+
+    def switch_maintenance_reason(switch):
+        if switch.id in maintenance_switches:
+            return {'maintenance'}
+        return set()
+
+    Switch.register_status_reason_func('maintenance', switch_maintenance_reason)
+
+
+New Topology REST Endpoints
+###########################
+
+The following set of new endpoints are intended to allow
+for previewing the affects of adding in interruptions.
+
+
+REST Endpoint ``GET /topology/v3/preview_interruption``
+-------------------------------------------------------
+
+For all known service providers,
+previews affected if the specified interruption occurs now.
+
+Expected Input:
+
+.. code-block:: json
+
+    {
+        "type": <interruption type str>,
+        "switches": [<switch id>],
+        "interfaces": [<interface id>],
+        "links": [<link id>],
+    }
+
+
+Expected Response:
+
+.. code-block:: json
+
+    {
+        <provider name str>: [<service name str>],
+    }
+
+Side Effects:
+
+- None
+
+Notes:
+
+- Should support multiple key-value pairs with the same key in the query parameters.
+
+
+Endpoint: ``GET <url for previewing>``
+--------------------------------------
+
+For the given service provider,
+previews services affected if the specified interruption occurs now.
+
+Expected Input:
+
+.. code-block:: json
```

Comment:
This thread https://github.com/kytos-ng/kytos/pull/356#discussion_r1157486183 is still unanswered. 

Can we only have service providers to not to have to filter for switches and links but only `interfaces`? 

Also, how are we going to specify which type of path? `path_type="primary|backup` ?

Reply:
To avoid confusion what primary|backup means, it's worth noting that for mef_eline has two major cases:

- Dynamic circuits: their current "primary" is always `"current_path"`, and the `"failover_path"` is what is considered a "backup"
- Static circuits: `mef_eline` uses `"primary_path"` and `"backup_path"`, while the `"current_path"` will follow whichever path is active.

Reply:
> This thread [#356 (comment)](https://github.com/kytos-ng/kytos/pull/356#discussion_r1157486183) is still unanswered.

So for `mef_eline`, the idea would be that rather than querying `/v2/evc` for the appropriate paths, we instead query a new endpoint `/v3/paths` which is an aggregation of all paths for all evcs.

> Also, how are we going to specify which type of path? `path_type="primary|backup` ?

As for how to differentiate between primary and secondary, just include it in the name of the new path. I don't think it would be necessary to have to make two separate queries when we know we are going to ask for them at the same time.

> Can we only have service providers to not to have to filter for switches and links but only `interfaces`?

Maybe? But, what's your logic behind that idea?

Reply:
> So for mef_eline, the idea would be that rather than querying /v2/evc for the appropriate paths, we instead query a new endpoint /v3/paths which is an aggregation of all paths for all evcs.

@Ktmi I had an impression that [on this thread](https://github.com/kytos-ng/kytos/pull/356#discussion_r1159884980), the client (`topology`) could perform two concurrent `GET <url for previewing>` for each `path_type`, that it'd be responsible for consolidating which service would be impacted based on their backup too. But for sure, that can also be delegated to each service provider to compute all of that too. Since you're considering this new `/v3/paths` can you give an example how it'd be potentially used including with payload inputs and response and `GET <url for previewing>` too, just so I can follow a full example?

Reply:
> > So for mef_eline, the idea would be that rather than querying /v2/evc for the appropriate paths, we instead query a new endpoint /v3/paths which is an aggregation of all paths for all evcs.
> 
> @Ktmi I had an impression that [on this thread](https://github.com/kytos-ng/kytos/pull/356#discussion_r1159884980), the client (`topology`) could perform two concurrent `GET <url for previewing>` for each `path_type`, that it'd be responsible for consolidating which service would be impacted based on their backup too. But for sure, that can also be delegated to each service provider to compute all of that too. Since you're considering this new `/v3/paths` can you give an example how it'd be potentially used including with payload inputs and response and `GET <url for previewing>` too, just so I can follow a full example?

This is what I'm thinking so far.

Inputs(Optional):

```json
{
    'type': <str>,
    'switches': [<str>],
    'interfaces': [<str>],
    'links': [<str>],
}
```

Expected output:

```json
[
    {
        'name': <evc_name>('(primary)'|'(backup)'|'(failover)'),
        'owner': <str or null>,
        'dynamic': <bool>,
        'path': <list[Links]>,
        'active': <bool>,
        'enabled': <bool>,
    },
]
```
Also note that active in this context should mean that the path is the current one in use by the EVC, and that enabled means that it is not being blocked by any interruptions. Now should we only care about enabled paths, active paths, or all paths when checking the affects of an interruption? Or do we extend `status_reason` into paths to check which interruptions they are already affected by, and filter out paths that are already affected by the interruption of the type we are checking for?

Reply:
> Or do we extend status_reason into paths to check which interruptions they are already affected by

@Ktmi, I think it makes a lot of sense to extend `status_reason` into paths too. That way, we still have the flexibility of exposing the resulting `status_reason`, and if there are multiple ones that way it could be nicely exposed too. 

> and filter out paths that are already affected by the interruption of the type we are checking for?
 
I think not filtering them out would make it easier to understand. It'd be nice to have the `status` and `status_reason` of each path in the output too, it could make interpretation easier. At the end of the day an user could be trying to preview an interruption, so if it's already being interrupted, then makes sense to return the full consolidated `status_reason`, right? (related to the discussion of the first paragraph of this comment)

> Now should we only care about enabled paths, active paths, or all paths when checking the affects of an interruption?

If we're discussing here about `GET <url for previewing>`, since it's trying to preview an interruption, I think displaying all paths would be great. By all paths, I mean both primary and backup and ideally abstracted away from the semantics of the service provider (if an EVC is dynamic or static, regardless of the name of the path it can have one that represents a primary and another backup or secondary, just so in the future when we support other NApp that provision a service we can follow the same pattern you're standardizing here for interruption preview). I think your idea to aggregate by paths sounds really great to help with this but with path aggregation on it too if possible, for instance, on `GET <url for previewing>` something along these lines (and maybe also including the path links if you think it's useful):

```
[{
  "id": <id>,
  "name": < evc_name >,
  "owner": < str or null > ,
  "paths": {
    "primary": {
      "status": <status>
      "status_reason": <status_reason>
    },
    "backup": {
      "status": status
      "status_reason": <status_reason>
    }
  }
}]
```

That way on `GET /topology/v3/preview_interruption`, it could optionally also further expose the paths with the reasons too. Wdyt? 

Regarding `/v3/paths` would we still need it? I have an impression the way you're proposing and modelling `GET /topology/v3/preview_interruption` it could internally uses the collection controller accordingly with new aggregations without a new endpoint. Let me know if I misunderstood something. 

Regarding, `'active': <bool>, 'enabled': <bool>,` I have an impression that'd be a bit redundant, since if an EVC isn't active or enabled it can't be further interrupted, can we remove it from the output or do you see any other case where it can be useful?

Reply:
We wouldn't need the new paths. Ultimately having paths separate opens up the possibility of testing a more distant scenario, where services can depend on other services, possibly across NApp boundaries, and interruptions will propagate recursively to help manage.

Reply:
Right let's go with `GET <url for previewing>` as it is then. If in the future, we also need to include the paths in the output shouldn't be that hard to augment the response, and that wouldn't break API.